src/charm.py

```diff
@@ -114,94 +120,110 @@ def __init__(self, *args):
-            None: If none of the conditionals match
+        try:
+            self._set_n2_information()
+        except ValueError:
+            event.add_status(BlockedStatus("Waiting for MetalLB to be enabled"))
 
-        """
-        if not self._amf_container.exists(path=CONFIG_DIR_PATH):
-            return WaitingStatus("Waiting for storage to be attached")
+        if self._csr_is_stored() and not self._get_current_provider_certificate():
+            event.add_status(WaitingStatus("Waiting for certificates to be stored"))
+            return
 
-        if not _get_pod_ip():
-            return WaitingStatus("Waiting for pod IP address to be available")
+        if not self._amf_service_is_running():
+            event.add_status(WaitingStatus("Waiting for SMF service to start"))
+            return
 
-        return None
+        event.add_status(ActiveStatus())
 
-    def _is_unit_in_non_active_status(self) -> Optional[StatusBase]:
-        """Evaluate and return the unit's current status, or None if it should be active.
+    def ready_to_configure(self) -> bool:
+        """Returns whether the preconditions are met to proceed with the configuration.
 
         Returns:
-            StatusBase: MaintenanceStatus/BlockedStatus/WaitingStatus
-            None: If none of the conditionals match
-
+            ready_to_configure: True if all conditions are met else False
         """
-        if not self.unit.is_leader():
-            # NOTE: In cases where leader status is lost before the charm is
-            # finished processing all teardown events, this prevents teardown
-            # event code from running. Luckily, for this charm, none of the
-            # teardown code is necessary to preform if we're removing the
-            # charm.
-            return BlockedStatus("Scaling is not implemented for this charm")
-
         if not self._amf_container.can_connect():
-            return MaintenanceStatus("Waiting for service to start")
+            return False
 
         if invalid_configs := self._get_invalid_configs():
-            return BlockedStatus(f"The following configurations are not valid: {invalid_configs}")
+            logger.info(f"The following configurations are not valid: {invalid_configs}")
+            return False
 
-        if relations_issues := self._check_relations():
-            return relations_issues
+        for relation in ["fiveg_nrf", "database", "certificates"]:
+            if not self._relation_created(relation):
+                logger.info(f"Waiting for {relation} relation")
+                return False
 
-        if container_issues := self._check_container_availability():
-            return container_issues
-        try:
-            self._set_n2_information()
-        except ValueError:
-            return BlockedStatus("Waiting for MetalLB to be enabled")
+        if not self._database_is_available():
+            logger.info("Waiting for the amf database to be available")
```

Comment:
How about we keep the logging together with the statuses? I think it will be easier to keep the messages in sync while refactoring if we keep the together.

Reply:
Yes, we can log the messages when a status change occurs. 
Done